src/python/py/models/builder.py

```diff
@@ -2613,6 +2642,48 @@ def make_layer(self, layer_id, layer):
             self.layernorm_attrs["last_layernorm"] = True
 
 
+class ChatGLMModel(Model):
+    def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options):
+        super().__init__(config, io_dtype, onnx_dtype, ep, cache_dir, extra_options)
+        # self.input_shapes["position_ids"] = [1]  # Note: This is optional and only needed if you want position_ids to be an int instead of a 2D tensor
+        self.layernorm_attrs["simple"] = True # RMSNorm in ChatGLM
+        self.rotemb_attrs["num_heads"] = self.num_attn_heads
+        self.rotemb_attrs["partial_rotary_factor"] = 0.5 # Line 755 of modeling_chatglm.py check self.rotary_pos_emb declaration
+        self.rotemb_attrs["rotary_embedding_dim"] = int(self.head_size * self.rotemb_attrs["partial_rotary_factor"])
+        self.rotemb_attrs["interleaved"] = 1
+        self.mlp_attrs["use_proj"], self.mlp_attrs["use_fc"] = True, False
+        self.attention_attrs["use_rotemb_in_attn"] = True
+        self.attention_attrs["use_packed_matmul"] = True
```

Comment:
The combination of these `attention_attrs` settings will cause the FP32 CUDA model to be created incorrectly. For the FP32 CUDA model, the `GroupQueryAttention` op does not have an implementation. The fallback is to use the `MultiHeadAttention` op, which does not fuse rotary embeddings within it and does not handle the extra `repeat_kv` operation needed when `num_attention_heads != num_key_value_heads`. Therefore, the FP32 CUDA model would need the following settings.
- `use_rotemb_in_attn = False`
- `use_packed_matmul = False`
- `op_type = "MultiHeadAttention"`

The subgraph for FP32 CUDA should look something like this.

```python
# MultiHeadAttention example:
#
#               root_input
#              /     |     \
#       Q_MatMul  K_MatMul  V_MatMul  4D causal mask  past_key  past_value
#           |        |         |            |            |           |
#       Q_Rotary  K_Rotary     |            +------------+-----------+
#           |        |         |                         |
#           |     Repeat_K  Repeat_V                     |
#           \        |        /                          |
#            MultiHeadAttention--------------------------+
#                    |
#                 O_MatMul
#                    |
#                  O_Add
```

Reply:
This relates to #880, for our use case, we would like `GroupQueryAttention` and our custom ep would have an implementation for it. I'm ok with changing the default behavior to adhere what is available in CPU/CUDA, as long as there is a way to override and generate with selected ops.